dwave/system/samplers/dwave_sampler.py

```diff
@@ -451,3 +460,78 @@ def _hook(computation):
                                             sort_labels=True)
 
     return _hook
+
+
+def failover(f):
+    @functools.wraps(f)
+    def wrapper(sampler, *args, **kwargs):
+        while True:
+            try:
+                return f(sampler, *args, **kwargs)
+            except SolverOfflineError:
+                pass
+
+            try:
+                # the requested features are saved on the client object, so
+                # we just need to request a new solver
+                sampler.solver = sampler.client.get_solver()
+
+                # delete the lazily-constructed attributes
+                try:
+                    del sampler._edgelist
+                except AttributeError:
+                    pass
+
+                try:
+                    del sampler._nodelist
+                except AttributeError:
+                    pass
+
+                try:
+                    del sampler._parameters
+                except AttributeError:
+                    pass
+
+                try:
+                    del sampler._properties
+                except AttributeError:
+                    pass
+
+            except SolverNotFoundError as err:
+                if sampler.retry_interval < 0:
+                    raise err
+
+                time.sleep(sampler.retry_interval)
+    return wrapper
+
+
+class DWaveFailoverSampler(DWaveSampler):
+    """Sampler that will recover from a QPU going offline by switching to a new
+    solver.
+
+    In the case that :exc:`~dwave.cloud.exceptions.SolverOfflineError` is raised
+    a new solver is requested.
+    If a :exc:`~dwave.cloud.exceptions.SolverNotFoundError` is also raised, the
+    sampler will wait `retry_interval` seconds before trying again.
+
+    Args:
+        retry_interval (number, optional, default=-1):
+            The amount of time (in seconds) to wait to poll for a solver in
+            the case that no solver is found. If `retry_interval` is negative
+            then it will instead propogate the `SolverNotFoundError` to the
+            user.
+
+        **config: See :class:`.DWaveSampler`.
+
+    """
+    def __init__(self, retry_interval=-1, **config):
+        super(DWaveFailoverSampler, self).__init__(**config)
+        self.retry_interval = retry_interval
+
+    @failover
```

Comment:
This looks like a leaky abstraction. I would at least have `failover` as a private member of `DWaveFailoverSampler`.